// pdfdataextract/dist/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PdfReferenceOutline = exports.PageNumberOutline = exports.UrlOutline = exports.Sort = exports.VerbosityLevel = void 0;
const pdf_1 = require("pdfjs-dist/legacy/build/pdf");
var VerbosityLevel;
(function (VerbosityLevel) {
    /**
     * logs all errors
     */
    VerbosityLevel[VerbosityLevel["ERRORS"] = pdf_1.VerbosityLevel.ERRORS] = "ERRORS";
    /**
     * logs all infos
     */
    VerbosityLevel[VerbosityLevel["INFOS"] = pdf_1.VerbosityLevel.INFOS] = "INFOS";
    /**
     * logs all warnings
     */
    VerbosityLevel[VerbosityLevel["WARNINGS"] = pdf_1.VerbosityLevel.WARNINGS] = "WARNINGS";
})(VerbosityLevel = exports.VerbosityLevel || (exports.VerbosityLevel = {}));
var Sort;
(function (Sort) {
    /**
     * if it should sort ascending
     */
    Sort[Sort["ASC"] = 0] = "ASC";
    /**
     * if it should sort descending
     */
    Sort[Sort["DESC"] = 1] = "DESC";
})(Sort = exports.Sort || (exports.Sort = {}));
/**
 * the outline which includes a url
 */
class UrlOutline {
    /**
     * @param {string} title - the title
     * @param {string} url - the url to which the outline points
     * @param {boolean} absolute - if the url is absolute
     * @param {Outline[]} [childs] - the childrens
     */
    constructor(title, url, absolute, childs) {
        this.title = title;
        this.url = url;
        this.absolute = absolute;
        this.childs = childs;
    }
}
exports.UrlOutline = UrlOutline;
/**
 * the outline which includes a page number
 */
class PageNumberOutline {
    /**
     * @param {string} title - the title
     * @param {number} page - the page number to which the outline points
     * @param {Outline[]} [childs] - the childrens
     */
    constructor(title, page, childs) {
        this.title = title;
        this.page = page;
        this.childs = childs;
    }
}
exports.PageNumberOutline = PageNumberOutline;
/**
 * the outline which includes a reference to another pdf
 */
class PdfReferenceOutline {
    /**
     * @param {string} title - the title
     * @param {string} url - the url to which the outline points
     * @param {number} page - the remote page number to which the outline points
     * @param {Outline[]} [childs] - the childrens
     */
    constructor(title, url, page, childs) {
        this.title = title;
        this.url = url;
        this.page = page;
        this.childs = childs;
    }
}
exports.PdfReferenceOutline = PdfReferenceOutline;
